### ISSUE TITLE
Statistics on individual trajectory durations

### ISSUE THREAD

**Issue by AmirHKiani98** [I24-MOTION/I24M_improvement_tracker#6]
Hi,
I was working with the dataset but something got my attention. I tried to see what is the timespan for each trajectory existed in the dataset. I got the following:
![image](https://github.com/I24-MOTION/I24M_improvement_tracker/assets/74076077/980e3c67-5b38-44bd-a06b-668918046153)
According to the picture above, for most of the trajectories the timespan is less than 10 seconds. I'm thinking if I'm missing something. Doesn't each datapoint in the dataset refer to a whole trajectory of one vehicle? If so, shouldn't the timespan (duration) be longer? Any information would be sincerely appreciated.

Best regards,
Amir

**Comment by barbourww** [comment]
@AmirHKiani98 For this example, does the duration of the trajectories in question correspond to the number of points in the trajectory? For instance, if the trajectory duration (maximum value minus minimum value in the trajectory) is 3.2 seconds, we would expect 80 x/y values for the trajectory. I'm wondering if this is in fact a timestamp issue or if this is just fragmentation.
Also, can you please share a code snippet so we can recreate on our end?

**Comment by AmirHKiani98** [comment]
@barbourww As per the provided example by you, I can tell they correspond to the number of points. But still I'm wondering what's the speed limit on I24. I'm getting somevalues close to 105mph, which is somehow explainable. Some drivers might have such a speed at some points.  We have also some drivers who drive as slow as 40 mph and a few less than 10. Keep that in mind that these values are average speed of the vehicle through the whole trajectory.

As for the code snippet, I don't believe my code would be runnable on your end, as I preprocess your dataset and work on the polished dataset.

**Comment by junyi9** [comment]
@AmirHKiani98 An alternative method to assist us in examining the issue is by sharing a list of the top 10 trajectory IDs that you believe might have issues. This will enable us to investigate those specific trajectories.

**Comment by barbourww** [comment]
@AmirHKiani98 Without a code example I can't comment on the validity of any particular trajectories. We have updated our [data tutorial](https://github.com/I24-MOTION/data_tutorial) to incorporate some basic distributional statistics about trajectory distance, duration, and speed. There is definitely a level of fragmentation in this current version of the dataset that results in a large number of short distance/duration trajectory fragments. Due to the briefness of these fragments, they may contain noisy speed values that appear as outliers in the average. That being said, the speed limit of the roadway for the INCEPTION dataset is 70mph. Drivers frequently drive 80-85mph in the left lanes during free flow. During congestion, speeds dropping below 10mph is common and due to the aforementioned fragmentation, some of the trajectory average speeds may be that low.

**Comment by Yiru-Jiao** [comment]
Similar results here and I believe it's caused by fragmentation... It seems we need to wait for the subsequent versions where hopefully vehicles are consistently tracked across cameras. Here are some code that can help reproduce the problem of short duration.
```python
import pandas as pd
import numpy as np
import ijson
import matplotlib.pyplot as plt
import datetime
transtime = datetime.datetime.fromtimestamp
from tqdm import tqdm

# data reading function
def to_dataframe(file, num2retrieval):
    columns = ['track_id','frame_id','x','y']
    metadata_west = {'unique_id':[],'track_id':[],'direction':[],'length':[],'width':[],'height':[]}
    metadata_east = {'unique_id':[],'track_id':[],'direction':[],'length':[],'width':[],'height':[]}
    data_west = []
    data_east = []
    track_id_west = 0
    track_id_east = 0

    with open(file) as f:
        parser = ijson.items(f, 'item', use_float=True)
        for veh in tqdm(parser, total=num2retrieval):
            if track_id_west+track_id_east >= num2retrieval:
                break
            # create a dataframe for each vehicle
            df = pd.DataFrame(np.zeros((len(veh['timestamp']),len(columns))), columns=columns)
            # fill in metadata
            if veh["direction"] == -1:
                metadata = metadata_west
                data = data_west
                track_id = track_id_west
                track_id_west += 1
            elif veh["direction"] == 1:
                metadata = metadata_east
                data = data_east
                track_id = track_id_east
                track_id_east += 1
            keys = ['unique_id','track_id','direction','length','width','height']
            values = [veh['_id']['$oid'], track_id, veh['direction'], veh['length'], veh['width'], veh['height']]
            for key, value in zip(keys, values):
                metadata[key].append(value)
            # fill in vehicle movement data
            df['track_id'] = track_id
            df['frame_id'] = np.array([int(transtime(t).strftime('%H%M%S%f')[:-4]) for t in veh['timestamp']])
            df['x'] = np.array(veh['x_position'])
            df['y'] = np.array(veh['y_position'])
            # df['vx'] = np.gradient(df['x'], df['frame_id']/100)
            # df['vy'] = np.gradient(df['y'], df['frame_id']/100)
            # df['speed'] = np.sqrt(df['vx']**2 + df['vy']**2)
            data.append(df)
            
    metadata_west = pd.DataFrame(metadata_west)
    data_west = pd.concat(data_west).reset_index(drop=True)
    metadata_east = pd.DataFrame(metadata_east)
    data_east = pd.concat(data_east).reset_index(drop=True)

    return metadata_west, data_west, metadata_east, data_east

# read data
metadata_west, data_west, metadata_east, data_east = to_dataframe(path_rawdata + '637c399add50d54aa5af0cf4__post2.json', 1000)

# duration distribution
fig, axes = plt.subplots(1,2,figsize=(8,3))
bins = np.linspace(0,30,31)
axes[0].hist(data_east.groupby('track_id')['frame_id'].count()/25, bins=bins)
axes[1].hist(data_west.groupby('track_id')['frame_id'].count()/25, bins=bins)
axes[0].set_title('Eastbound')
axes[1].set_title('Westbound')
axes[0].set_xlabel('Duration (s)')
axes[1].set_xlabel('Duration (s)')
axes[0].set_ylabel('Frequency')
```
In addition I noticed that some vehicles that are located in the area of westbound lanes (y>0) are indicated as along the direction of eastbound (and vice versa). My guess is that `direction` was determined by trajectory movement direction instead of `y` coordinates. Below the code can find such examples.
```python
print(metadata_west[metadata_west['track_id'].isin(data_west[data_west['y']<0]['track_id'].unique())])
print(metadata_east[metadata_east['track_id'].isin(data_east[data_east['y']>0]['track_id'].unique())])
```

**Comment by aakabiri** [comment]
I actually have the same issue. I read one day of data (Nov 28) and found that there are more than 700,000 unique vehicle IDs, and the median duration of observation for trajectories (last_timestamp - first_timestamp) is around 12 seconds. My question is does the data track a single vehicle along the road across different cameras on the road? Or do different cameras catch a specific vehicle by multiple vehicle IDs?

**Comment by junyi9** [comment]
@Yiru-Jiao @aakabiri Yes, it is challenging to track one vehicle across multiple cameras. Our recent paper detailing the challenges is highly recommended.

>Gloudemans, Derek, et al. "So you think you can track?." Proceedings of the IEEE/CVF Winter Conference on Applications of Computer Vision. 2024. [paper link](https://openaccess.thecvf.com/content/WACV2024/papers/Gloudemans_So_You_Think_You_Can_Track_WACV_2024_paper.pdf)
